### PR TITLE
Add new function to enable/disable preview of map to lua

### PR DESCRIPTION
**Pull request by ottml**
We need lua functionality to have possibility to disable the map preview for missions. So the user can not see the map in advance.

previewdisabled
previewenabled

### CODE REVIEW

File: libs/s25main/lua/LuaInterfaceSettings.cpp [hunk 1]
```diff
@@ -237,3 +237,15 @@ std::vector<AddonId> LuaInterfaceSettings::GetAllowedAddons()
...
+
+bool LuaInterfaceSettings::IsMapPreviewEnabled()
+{
+    kaguya::LuaRef getMapPreviewEnabled = lua["isMapPreviewEnabled"];
+    if(getMapPreviewEnabled.type() == LUA_TFUNCTION)
+    {
+        kaguya::LuaRef previewEnabled = getMapPreviewEnabled();
+        if(previewEnabled.typeTest<bool>())
+            return previewEnabled;
+    }
```
Comment (Flamefire) [inline libs/s25main/lua/LuaInterfaceSettings.cpp:249]: Simpler version similar to https://github.com/Return-To-The-Roots/s25client/blob/605bd80f5cbb7c123dc5ba581124baae70ec783c/libs/s25main/lua/LuaInterfaceSettings.cpp#L180

```suggestion
        return getMapPreviewEnabled.call<bool>();
```

And maybe rename to `func` as there is already a naming-mix between "get" and "is"

File: tests/s25Main/lua/testLuaSettings.cpp [hunk 2]
```diff
@@ -149,6 +149,10 @@ BOOST_AUTO_TEST_CASE(Events)
     BOOST_TEST_REQUIRE(allowedAddons[0] == AddonId::LIMIT_CATAPULTS);
     BOOST_TEST_REQUIRE(allowedAddons[1] == AddonId::CHARBURNER);
     BOOST_TEST_REQUIRE(allowedAddons[2] == AddonId::TRADE);
+
+    executeLua("function isMapPreviewEnabled()\n  return { false }\nend");
+    BOOST_TEST_REQUIRE(lua.IsMapPreviewEnabled());
```
Comment (Flamefire) [inline tests/s25Main/lua/testLuaSettings.cpp:154]: This is the opposite, isn't it?

File: tests/testData/maps/LuaFunctions.lua [hunk 3]
```diff
@@ -20,6 +20,10 @@ function getAllowedChanges()
 	return {general=true, addonsAll=false, addonsSome=true, swapping=false, playerState = not isSinglePlayer, ownNation = true, ownColor=true, ownTeam=false, aiNation = false, aiColor=false, aiTeam=false}
 end
 
+function isMapPreviewEnabled()
```
Comment (Flamefire) [inline tests/testData/maps/LuaFunctions.lua:23]: Any reason to add this here?

Comment (ottml) [reply tests/testData/maps/LuaFunctions.lua:23]: I added it because other settings are there too.

**Comment by Flow86** [comment]
can we perhaps instead of not showing any map if preview is disabled, showing some "Hidden" map view, don't know if we can generate a preview of something like that?

**Review by Flamefire — DISMISSED** [review]

**Comment by Flamefire** [comment]
> can we perhaps instead of not showing any map if preview is disabled, showing some "Hidden" map view, don't know if we can generate a preview of something like that?

Maybe some placeholder graphics. Not sure if it is worth the effort. But if e.g. @Spikeone wants to provide an image @ottml likes I'm fine with it. Otherwise I'd just keep the current state of the PR

**Comment by Spikeone** [comment]
@Flamefire I'd to this without a graphic. Can't we just add a gray area (like for player colors) and either use a scaled up '?' or simply write "no preview" or something alike in that area?

**Comment by Flow86** [comment]
> Can't we just add a gray area (like for player colors) and either use a scaled up '?' or simply write "no preview" or something alike in that area? 

thats a good idea

**Comment by ottml** [comment]
disabledView
disabledView2

File: libs/s25main/desktops/dskGameLobby.cpp [hunk 4]
```diff
@@ -231,27 +231,36 @@ dskGameLobby::dskGameLobby(ServerType serverType, std::shared_ptr<GameLobby> gam
...
-        libsiedler2::Archiv mapArchiv;
-        // Karteninformationen laden
-        if(int ec = libsiedler2::loader::LoadMAP(GAMECLIENT.GetMapPath(), mapArchiv))
+        bool isMapPreviewEnabled = !lua || lua->IsMapPreviewEnabled();
+        if(!isMapPreviewEnabled)
         {
-            WINDOWMANAGER.ShowAfterSwitch(
-              std::make_unique<iwMsgbox>(_("Error"), _("Could not load map:\n") + libsiedler2::getErrorString(ec), this,
-                                         MsgboxButton::Ok, MsgboxIcon::ExclamationRed, 0));
+            AddColorDeepening(70, DrawPoint(560, 40), Extent(220, 220), TextureColor::Grey, 0);
```
[the remark this answers is not in the document]
Comment (ottml) [reply libs/s25main/desktops/dskGameLobby.cpp]: Yes. Good idea

File: tests/s25Main/lua/testLuaSettings.cpp [hunk 5]
```diff
@@ -149,6 +149,10 @@ BOOST_AUTO_TEST_CASE(Events)
     BOOST_TEST_REQUIRE(allowedAddons[0] == AddonId::LIMIT_CATAPULTS);
     BOOST_TEST_REQUIRE(allowedAddons[1] == AddonId::CHARBURNER);
     BOOST_TEST_REQUIRE(allowedAddons[2] == AddonId::TRADE);
+
+    executeLua("function isMapPreviewEnabled()\n  return false\nend");
+    BOOST_TEST_REQUIRE(!lua.IsMapPreviewEnabled());
```
Comment (Flamefire) [inline tests/s25Main/lua/testLuaSettings.cpp:154]: Also check the default:

```suggestion
    BOOST_TEST(lua.IsMapPreviewEnabled()); // Not defined in lua -> Default to enabled
    executeLua("function isMapPreviewEnabled()\n  return false\nend");
    BOOST_TEST(!lua.IsMapPreviewEnabled());
    executeLua("function isMapPreviewEnabled()\n  return true\nend");
    BOOST_TEST(lua.IsMapPreviewEnabled());
```

(Yes we erroneously used `BOOST_TEST_REQUIRE` for no good reason around this)

**Review by Flamefire — CHANGES_REQUESTED** [review]
Looks good, so just a minor improvement on the test.

**Review by Flamefire — APPROVED** [review]

**Comment by Spikeone** [comment]
hey @ottml just asking since I've seen something alike in the campagin.lua - wouldn't it be great to set some kind of preview graphic - so one could for example set the SETUP810.LBM in order to hint, that this is winter map.

**Pull request merged**